api/maven-api-core/src/main/java/org/apache/maven/api/Artifact.java

```diff
@@ -105,6 +105,14 @@ default String key() {
      */
     boolean isSnapshot();
 
+    /**
+     * The artifact properties.
+     *
+     * @return the artifact properties, never {@code null}
+     */
+    @Nonnull
+    ArtifactProperties getArtifactProperties();
```

Comment:
Exactly, happens similarly as in Resolver:
* IF "ad-hoc" resolution happened (i.e. plugin used resolve service) it is _plugin that knows why it did that_
* IF artifact entered via project (as dependency) then you have dependency -> type -> artifact.

Reply:
I disagree here: Dependency is what you have in POMs (either project or transitive deps). So, when you resolve, you end up with bunch of `Artifact`s, NOT `Dependency`ies. Basically, that is why this is needed on `Artifact` and not `Dependency` (where is also needed, as properties from type should and could be still augmented).

Reply:
Or in other words: see this PR https://github.com/apache/maven-resolver/pull/411
Basically Resolver can be now asked to:
* create Artifact as `new Artifact("g:a:v", artifactTypeRegistry.get("annotation-processor"))`.
* the to resolve it to have it's full transitive hull
* and you again end up with "bunch of Artifacts"

No dependency in scope, but you are STILL interested in Artifact properties....

Or similarly for ANY artifact, as if artifact you are resolving may have type="modular-jar", then the needed info will be still present in artifact.properties. Again, no dependencies involved, etc.

Am pretty much confident that this change is correct.

Reply:
> Or in other words: see this PR [apache/maven-resolver#411](https://github.com/apache/maven-resolver/pull/411) Basically Resolver can be now asked to:
> 
> * create Artifact as `new Artifact("g:a:v", artifactTypeRegistry.get("annotation-processor"))`.
> * the to resolve it to have it's full transitive hull
> * and you again end up with "bunch of Artifacts"

> 
> No dependency in scope, but you are STILL interested in Artifact properties....

I don't think that's right.  You're looking at the resolver API, not the Maven API here.

> Or similarly for ANY artifact, as if artifact you are resolving may have type="modular-jar", then the needed info will be still present in artifact.properties. Again, no dependencies involved, etc.

There's no `Type` on the `Artifact` ...

> Am pretty much confident that this change is correct.

I think it maps to the resolver api and model, but definitely not to the Maven 4 API, which as `Artifact`, `ArtifactCoordinate`, `Dependency` and `DependencyCoordinate`, whereas resolver as `Artifact` only...

Reply:
No, `Type` is a property of `Dependency`. What happens is that is just an indirection of `Type` -> `properties`. If you think about, EVERY Artifact (sans root) have type (as they enter scope as dependencies of root).

And yes, am looking at Resolver API as it is done well and works. IMO we should really just pull a thin layer on top of it for start, that we can refine later. But in general, this is how things should work, really.

Reply:
> No, `Type` is a property of `Dependency`, no Artifact. What happens is that is just an indirection of `Type` -> `properties`. If you think about, EVERY Artifact (sans root) have type (as they enter scope as dependencies of root).

That's wrong. Artifacts can be resolved without any Type / Dependency at all using the `ArtifactResolver`:
https://github.com/apache/maven/blob/2788849c37ed4c425eb9532c5e20b17f85cbe396/api/maven-api-core/src/main/java/org/apache/maven/api/services/ArtifactResolver.java

That's the whole point.  ArtifactCoordinate is used to obtain an Artifact.  A DependencyCoordinate is used to obtain a Dependency.  `DependencyCoordinate` contains `Type` and `Scope`.

If we need some properties on `Artifact`, that's fine, we can define `ArtifactProperties` and `DependencyProperties` which extends `ArtifactProperties`. The artifact properties should contain only properties that are related to an artifact, not its usage (type, scope, etc...).  But my understanding is that there's no properties at the moment.  The only one in the resolver is the path of the file, which is obtained from the `ArtifactManager`.

> And yes, am looking at Resolver API as it is done well and works. IMO we should really just pull a thin layer on top of it for start, that we can refine later. But in general, this is how things should work, really.

Reply:
We need open properties (my 2cts would be it is `Map<String, Object>`, no need to type it stronger) to enable a (custom) type - I include the "module-jar" kind of types, ie all but extensions types - to complete the meta of a dependency - the old artifact handler somehow which is propagated to properties.
Having or not a wrapper is a style choice but we already have some properties (is added to classpath etc) no?

Reply:
Sorry, used wrong lingo. You are right "artifact resolve" is really "just download that single file for that artifact, no deps involved". I was talking about `collect+resolve` that is the main use case of Resolver, as there, all the artifacts you end up with (resolved) are dependencies except the root, but that one as well may have properties if you used `ArtifactType` to construct it.

Reply:
> Sorry, used wrong lingo. You are right "artifact resolve" is really "just download that single file for that artifact, no deps involved". I was talking about `collect+resolve` that is the main use case of Resolver, as there, all the artifacts you end up with (resolved) are dependencies except the root, but that one as well may have properties if you used `ArtifactType` to construct it.

That's exactly why I think we should keep things split more clearly between Artifact and Dependency.  We have artifacts which are related to repositories and dependencies for... dependencies.  Different words for different meanings.  We should make that more clear rather than blur things...

Reply:
> ...we already have some properties (is added to classpath etc) no?

@rmannibucau 

That method is one of the main sources of confusion :smile: This is what happens:
* Maven2 had `ArtifactHandler#isAddedToClasspath()` which meant exactly that: "is artifact to be added to classpath" as Maven2 was aware only of classpath, nothing else.
* in Maven3 resolver came, and introduced this flag: https://github.com/apache/maven-resolver/blob/master/maven-resolver-api/src/main/java/org/eclipse/aether/artifact/ArtifactProperties.java#L54-L62 (note the javadoc nor any string literal **does not mention classpath or anything**).
* In development of Maven3 early, at one point these two were "shortcut into one": https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/RepositoryUtils.java#L247-L268
* so essentially the "classpath" notion in core coming from Maven2 and "buildpath" abstraction, intentionally defined in Resolver as such (new in Maven3), was blurred to mean "same thing", which was obviously not the intent.

Typical example: the POM packaging, see definition here:
https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/internal/impl/types/PomTypeProvider.java

Long time ago, before BOMs were a thing, people used it to "group dependencies" (see for example this blog: https://blog.sonatype.com/2009/10/maven-tips-and-tricks-grouping-dependencies/).

In essence, you create a POM that has your grouped dependencies and deploy that. So you have:
```
G:A:pom:V
  +- D1
  +- D2
  +- D3
```

On consuming side, when you declare in your project dependency (yes, a dependency, not `import` as today you do with BOMs) like `G:A:pom:V`, hence as:

```
<dependency>
  <groupId>G</groupId>
  <artifactId>A</artifactId>
  <version>V</version>
  <type>pom</type>
</dependency>
```

On resolving (for @gnodet collect+resolve) this project "compile" build path scope, assuming is the only dependency you have, you would end up with this list of artifacts: [D1, D2, D3] despite resolution started as `P` (root) where P had 1 dependency `G:A:pom:V`. 

Simply put, the Type definition (lack of "constitutes build path") causes that intermediate "grouping POM" be completely omitted (left out) from the results, as it is not part of the build path.

Reply:
Side note: the long time ago pattern of doing pom for deps is still accurate today and bom created more mess than helped there - but unrelated to the PR, just wanted to highlight this is still used for goods and not legacy.

Ok but it means we need a dependency handler as we have an artifact handler (think it is a bit complex).
Ultimately, do we need dependency outside the model (literally <dependency>)  or can it not just be artifacts?
I understand what you mean but it seems like a lot of work for not much gain, no? Can't we merge it somehow in new api?

Reply:
> Side note: the long time ago pattern of doing pom for deps is still accurate today and bom created more mess than helped there - but unrelated to the PR, just wanted to highlight this is still used for goods and not legacy.
> 
> Ok but it means we need a dependency handler as we have an artifact handler (think it is a bit complex). Ultimately, do we need dependency outside the model (literally ) or can it not just be artifacts? I understand what you mean but it seems like a lot of work for not much gain, no? Can't we merge it somehow in new api?

There's no ArtifactHandler in the v4 api.  It's called `Type` and it represents the usage of an Artifact as a Dependency.  There's nothing to handle afaik on an Artifact: it's just a pointer to a file in a repository.  How it's used needs to be taken care of and that's done in the `Dependency` which is associated to a `Type` and `Scope` and `DependencyProperties`.  If we move things to the Artifact or merge Artifact and Dependency, I think that's a source of confusion.

Reply:
Resolver also encodes "type" in Artifact#properties. It also exposed since eons ctors for DefaultArtifact where you could pass in ArtifactType. The type you declare (i.e. if in project, you have packaging -> type) or if "ad-hoc" by root dependency node type (where callar knows what it does).

Or in short:
* artifact is a file decorated with coordinates and properties
* dependency is an artifact having extras like scope, type, optionality using which you can "derive" an artifact as above
* true, in Resolver "dependency" is ONLY a thing that comes from POM (ArtifactDescriptor) but it is being distilled to artifact (as in first bullet).

Basically "everything is an artifact", and "dependency is an artifact that enters the game as dependency". This is how Resolver abstracts. 

Doing collect+resolve operation and ending up with list of sole Dependencies is what you say?

When you collect+resolve, you are interested in all, so even root (again, this is for "ad-hoc" operations, like javadoc may tell: "resolve me this taglet and its dependencies", in that case, javadoc plugin KNOWS root is taglet, but will still need to handle the rest of transitive hull, and place them or use that as needed).

---

@gnodet if I understand you correctly, sans "root", you consider _everything else_ (the transitive hull) as "dependency"? Which kinda makes sense, as they DID enter the result collection as "dependencies of root"?

Reply:
> Resolver also encodes "type" in Artifact#properties. It also exposed since eons ctors for DefaultArtifact where you could pass in ArtifactType. The type you declare (i.e. if in project, you have packaging -> type) or if "ad-hoc" by root dependency node type (where callar knows what it does).
> 
> Or in short:
> 
> * artifact is a file decorated with coordinates and properties
> * dependency is an artifact having extras like scope, type, optionality using which you can "derive" an artifact as above
> * true, in Resolver "dependency" is ONLY a thing that comes from POM (ArtifactDescriptor) but it is being distilled to artifact (as in second bullet -> first bullet).
> 
> Basically "everything is an artifact", and "dependency is an artifact that enters the game as dependency". This is how Resolver abstracts.

Yes, and I think that's where some confusion comes from.  Everything is an artifact means that when you have an Artifact object, you don't really know what it is: it can be 5 different things:
 * a pointer to a file (with a version range): ArtifactCoordinate
 * a resolved artifact in a repository (a definite version): Artifact
 * a pointer to a dependency (version range + scope + type)
 * a resolved dependency: the resolver actually has `org.eclipse.aether.graph.Dependency` and that's what the v4 api wraps
 * an artifact generated by the project

> Doing collect+resolve operation and ending up with list of sole Dependencies is what you say?

This is also a confusing part.  It's conceptually a 3 step process: collect + flatten + resolve.  The collect grabs the input and computes the graph of Node.  Flattens will output a *list* of Node.  Then the resolution step becomes the same as the _artifact resolution_  because each node can be turned into an artifact coordinate, and we then just download the actual artifacts.

In the v4 api, we have `ArtifactResolver` which gives an `Artifact`.   The `DependencyCollector` returns the graph with `Node`s.   The `DependencyResolver` performs both `flatten` which gives an ordered list of `Node`s, and `resolve` which returns an ordered `Map<Dependency, Path>` which I think is what is needed to compute the build paths.

> When you collect+resolve, you are interested in all, so even root (again, this is for "ad-hoc" operations, like javadoc may tell: "resolve me this taglet and its dependencies", in that case, javadoc plugin KNOWS root is taglet, but will still need to handle the rest of transitive hull, and place them or use that as needed).
> 
> @gnodet if I understand you correctly, sans "root", you consider _everything else_ (the transitive hull) as "dependency"? Which kinda makes sense, as they DID enter the result collection as "dependencies of root"?

Yes, everything is a dependency.  It's only the root which can be a project or artifact, but when you collect the dependencies for a project or artifact, you actually resolve all its *dependencies*.

I think there may be a small piece missing, which is the ability to access the graph from a dependency, i.e. adding a `Node getNode()` on the `org.apache.maven.api.Dependency` to get back to the graph.

At some point, I did envision to call it `ResolvedDependency` which would avoid the confusion with the _model dependency_, which is is translated into a `DependencyCoordinate`, the collected / flattened / resolved.

To sum up: the v3 api lacks a bunch of things, that's why plugins have to use the resolver to get fine grain access to the resolution output.  The v4 api aims to expose those bits, but those aren't really new, the `org.apache.maven.api.Dependency` simply wraps `org.eclipse.aether.graph.Dependency`.  That was clearly a missing piece in the v3 api.